### ISSUE TITLE
Improve Commit Messages

### ISSUE THREAD

**Issue by simlu** [dependabot/feedback#443]
Currently dependabot uses multiple different commit message syntaxes. Since the most recent commits are used to determine which style to use, the style can change over time and it is not always intuitive when and why it does.

This is obviously not ideal when e.g. setting up tooling that relies on commit message style. This issue aims to find a more general solution that satisfies more requirements than the current one and allows to "pin" the commit syntax.

-------------

**Suggestion**

Since dependabot already has a configuration file, a template string might work nicely. I.e.

```yml
commitSyntax: 'chore($scope): $msgShort\n$msgLong'
```

Could also support a syntax per scope, but that might be overkill.

**Comment by greysteil** [comment]
Related to https://github.com/dependabot/feedback/issues/186.

The full logic for building Dependabot's commit messages is [here](https://github.com/dependabot/dependabot-core/blob/master/common/lib/dependabot/pull_request_creator/message_builder.rb).

I agree we should make this configurable from Dependabot's config files. My hunch is that a menu of options (angular, ESLint, gitmoji, perhaps another I don't know of) would be the way to go, rather than making it fully configurable. The complaint I hear most often about semantic/conventional commits is that they make projects less accessible to new contributors as they're another convention to learn. If Dependabot can help folks converge on a standard within those communities I'd like to do that, and I think it would lead to a simpler interface than allowing full configuration.

**Comment by simlu** [comment]
A long time ago I've started with automation by setting up travis for a big java project that I was maintaining. Releases were few and far between. So everytime I went to "magically" do a release via travis it was broken and I had to fix it. I got frustrated, since it didn't actually save any time.

So I'm looking at this from the perspective of someone who embraces automation in every part of their development pipeline. I set up automation because I don't want to remember how things work.

Conventions change and what might be a good convention today might be outdated tomorrow. Everything that "magically" changes over time is bad because it will break your automation flow that you spend so much time setting up).

Dependabot is now in charge of defining when a new convention is "big enough that it should be added" (you already mentioned that there are two different angular conventions, what if a third one comes out). Also how would you handle subtle changes in existing conventions so that existing setups are non breaking (angular deprecates both conventions and comes out with a better revised one that everyone should use)?

I'm all for educating the user, but I don't think it's dependabots place to do that. Other than in a config generation tool or where the config is displayed in the dependabot GUI - maybe display a warning if the string doesn't match a known convention.

tl;dr Pick your battles, I don't think dependabot should do too much magic behind the scenes here. Many users won't care about the commit message style and for the other ones it's easy enough to find the correct template string from a readme and put it into the config.

**Comment by simlu** [comment]
After writing this I realised that there is another point here: From the perspective of doing dependabot configuration through the GUI an options menu might be preferred. When you set up a configuration file, you want full control, because you're already going into the lowest level.

Possibly solution: Since dependbot is already doing prs, I would love if the gui was just a way of editing the configuration file via prs. That way you would get the best of both worlds (I realise this would be a lot of work, but it would overcome both concerns here).

You could also keep things separate and have an options menu in the GUI and a `commitSyntax` in the  config file. That would be more in line with you currently have.

**Comment by simlu** [comment]
Let's try to answer some questions: If you were to go with a defined list of options, what would those be in more details? How would you handle the two angular styles? How would you handle different PR titles? A dropdown list as well?

**Comment by richardson-trevor** [comment]
Chiming in here with a use-case. We also use semantic-release, and we like to let it go all the way to production if tests pass. 

If we have a dev dependency (primarily test libs, etc.), we prefer that they are `chore`, whereas we want production dependencies to become `fix` . This way we cut a new release and get it deployed automatically for things like security updates in a production dep. 

So far, we've handled this by writing a semantic release commit analyzer and we just read Dependabot's commits and determine if it is `chore(deps)` or `chore(deps-dev)` and changes semantic-release's behavior accordingly.

**Comment by simlu** [comment]
That's exactly our use case. We handle it similarly and enforce the `.releaserc.json` file as defined [here](https://github.com/blackflux/robo-config-plugin/blob/master/src/plugin/tasks/semantic-release/commit-conventions.json) through [robo-config](https://github.com/blackflux/robo-config).

Which then results in [this](https://github.com/blackflux/robo-config-plugin/blob/master/.releaserc.json). You can see that we needed to add both conventions dependabot uses.

Yet another configuration file and but since we manage them centrally not a big deal. Having said that it would be great if dependabot could eliminate the need for that config file altogether.

**Comment by simlu** [comment]
@richardson-trevor I've looked at the dependabot code again. There more branching going on ([Reference](https://github.com/dependabot/dependabot-core/blob/master/common/lib/dependabot/pull_request_creator/message_builder.rb#L217)).

Hence our setups are actually pretty fragile when the commit conventions for the projects change. So solidification is much needed here on dependabots side.

**Comment by AdamVig** [comment]
I second this feature request. I need Dependabot to use a scope of `fix` instead of `chore`.

**Comment by dmsergeevN26** [comment]
This request is extremely important for us and the lack of this functionality is preventing us from using Dependabot. We need to be able to configure the prefix from a config file.

**Comment by greysteil** [comment]
@AdamVig @dmsergeevN26 understood. We are still a tiny team and have our work cut out integrating Dependabot directly into GitHub. This feature needs proper thought and we don't want to rush it, so I can't promise it in the short term.

**Comment by greysteil** [comment]
Everyone on this thread: I'd love your eyes on https://github.com/dependabot/dependabot-core/pull/1227. Will it give you what you need?

**Comment by dmsergeevN26** [comment]
Thanks. It would give us a way to configure the commit message prefix, but not the branch name

**Comment by greysteil** [comment]
Configuring branch names sounds like a separate issue to me - do you want to open a new issue to discus that?

**Comment by dmsergeevN26** [comment]
Sure

**Comment by simlu** [comment]
It would make the semantic-release use case mentioned several times before possible and resilient. However it will still require a custom releaserc config. Allowing for different prefixes would remove that: ie if we could say dev deps are a chore and deps are a fix

But much, much better than what is currently there! So definitely a good pr! Thank you!

**Comment by AdamVig** [comment]
I second the use case mentioned in the previous comment: at my company, development dependency updates should always be `chore` and regular dependency updates should be `fix`.

With that said, the proposed configuration file change is a step in the right direction. Thanks!

**Comment by greysteil** [comment]
Fixed by https://github.com/dependabot/dependabot-core/pull/1227. For anyone who wants to customise the prefix based on whether it's a production / development dependency, can you open a separate issue, please? :-)

**Comment by dmsergeevN26** [comment]
@greysteil Opened https://github.com/dependabot/feedback/issues/545 for the branch name issue

**Comment by richardson-trevor** [comment]
The new issue for configuring unique commit messages per dev/prod deps can be found in https://github.com/dependabot/feedback/issues/551.

**Comment by simlu** [comment]
@greysteil Thank you very much! Terrific addition!
